test/utils/executor.hpp

```diff
@@ -122,18 +98,31 @@ inline void init_executor(std::shared_ptr<gko::ReferenceExecutor> ref,
     if (gko::HipExecutor::get_num_devices() == 0) {
         throw std::runtime_error{"No suitable HIP devices"};
     }
-    exec = gko::HipExecutor::create(
-        0, ref, std::make_shared<gko::HipAllocator>(), stream);
+    exec =
+        gko::HipExecutor::create(ResourceEnvironment::hip_device_id, ref,
+                                 std::make_shared<gko::HipAllocator>(), stream);
 }
 
 
 inline void init_executor(std::shared_ptr<gko::ReferenceExecutor> ref,
                           std::shared_ptr<gko::DpcppExecutor>& exec)
 {
     if (gko::DpcppExecutor::get_num_devices("gpu") > 0) {
-        exec = gko::DpcppExecutor::create(0, ref, "gpu");
+        exec = gko::DpcppExecutor::create(ResourceEnvironment::sycl_device_id,
+                                          ref, "gpu");
     } else if (gko::DpcppExecutor::get_num_devices("cpu") > 0) {
-        exec = gko::DpcppExecutor::create(0, ref, "cpu");
+        if (ResourceEnvironment::sycl_cpu_id < 0) {
+            exec = gko::DpcppExecutor::create(0, ref, "cpu");
+        } else {
+            auto devices = create_cpu_sub_devices();
+            if (ResourceEnvironment::sycl_cpu_id >= devices.size()) {
+                throw gko::OutOfBoundsError(__FILE__, __LINE__,
+                                            ResourceEnvironment::sycl_cpu_id,
+                                            devices.size());
+            }
+            exec = gko::DpcppExecutor::create(
+                *devices[ResourceEnvironment::sycl_cpu_id], ref);
```

Comment:
Alternatively: 
```suggestion
            exec = gko::DpcppExecutor::create(
                *devices[ResourceEnvironment::sycl_cpu_id % devices.size()], ref);
```
which always works, but potentially oversubscribes the CPU

Reply:
I think we should over-subscribe the resources a bit anyway. It might help us with load balancing issues.
But maybe that could be done at the runner level.